### DISCUSSION TITLE
How to integrate my library into the firmware?

### DISCUSSION THREAD

**Discussion by xkhainguyen** [bitcraze/discussions#734] [Q&A] [answered]
Hello,
I can develop an LQR controller as an app layer (without an additional library/sub-directory). However, when I try to integrate my library, it is quite tricky with Kbuild and make files.
My previous topic is [here](https://github.com/orgs/bitcraze/discussions/657)
My library looks like this:
```
.
└── my_controller/
    ├── CMakeLists.txt
    ├── src/
    │   ├── c1.h
    │   ├── c1.c
    │   ├── c2.h
    │   ├── c2.c
    │   ├── controller.h
    │   └── CMakeLists.txt
    └── external/
        └── linear_algebra_lib/
            ├── CMakeLists.txt
            └── src/
                ├── la1.h
                ├── la1.c
                ├── la2.h
                ├── la2.c
                ├── linear_algebra.h
                └── CMakeLists.txt
```
1/ Am I supposed to include all the output .o files in Kbuild? 
2/ Is there a better way to modify make files to be compatible with my CMakeLists.txt? How can I add my compile definition to the build system?
3/ How about other options, like building a static/dynamic library?
4/ What will happen if my controller computation exceeds its allowed frequency? Will it keep computing the previous stuff in the next call?

**Comment by krichardsson** [comment]
> 1/ Am I supposed to include all the output .o files in Kbuild?

I'm assuming your root directory (my_controller/) is somewhere in the firmware file tree. The easiest way is probably to look at existing code for a template, for instance the [controller directory](https://github.com/bitcraze/crazyflie-firmware/tree/master/src/modules/src/controller). You need to add a [KBuild file in the directory](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/Kbuild) specifying the .o files that are required. You also need to add your directory in the [KBuild file in the parent directory](https://github.com/bitcraze/crazyflie-firmware/blob/c4925370e3c8a35a79bc1e34f9799028751222f6/src/modules/src/Kbuild#L54).
To make includes work, you must [add your include directory to the make file](https://github.com/bitcraze/crazyflie-firmware/blob/c4925370e3c8a35a79bc1e34f9799028751222f6/Makefile#L55).

> 2/ Is there a better way to modify make files to be compatible with my CMakeLists.txt? How can I add my compile definition to the build system?

Not that I know about.

> 3/ How about other options, like building a static/dynamic library?

One option is to add your controller code as a git submodule. You can check out the libdw1000 in vendors as an example. You still have to add files to the KBuild file in vendors and header files to the make file. 

You can also add you code in src/lib, this is similar to the examples above. 

It is possible to build a static library as well, but I don't think we have any examples of that.

> 4/ What will happen if my controller computation exceeds its allowed frequency? Will it keep computing the previous stuff in the next call?

That will be pretty bad. Your controller is called from the stabilizer loop (runs at 1000Hz) and your controller will block the stabilizer loop until your function returns.

**Comment by xkhainguyen** [comment]
Thank you very much for your response. @krichardsson 

> I'm assuming your root directory (my_controller/) is somewhere in the firmware file tree. The easiest way is probably to look at existing code for a template, for instance the [controller directory](https://github.com/bitcraze/crazyflie-firmware/tree/master/src/modules/src/controller). You need to add a [KBuild file in the directory](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/Kbuild) specifying the .o files that are required. You also need to add your directory in the [KBuild file in the parent directory](https://github.com/bitcraze/crazyflie-firmware/blob/c4925370e3c8a35a79bc1e34f9799028751222f6/src/modules/src/Kbuild#L54).
To make includes work, you must [add your include directory to the make file](https://github.com/bitcraze/crazyflie-firmware/blob/c4925370e3c8a35a79bc1e34f9799028751222f6/Makefile#L55).

This is what I followed and it worked. I added it as an app. 

> Not that I know about.

I am able to add compile flags in Makefile within EXTRA_CFLAGS. 

> One option is to add your controller code as a git submodule. You can check out the libdw1000 in vendors as an example. You still have to add files to the KBuild file in vendors and header files to the make file. You can also add you code in src/lib, this is similar to the examples above. It is possible to build a static library as well, but I don't think we have any examples of that.

Yes submodule seems nice. I believe the correct way is to make the submodule source code highly compatible with the crazyflie firmware first, then put KBuild into .gitignore to avoid any modifications to the submodule source.

My another problem is that I want to use Eigen. I am able to compile it for another STM32 with g++ (into C code). However, I cannot figure out how to do that with crazyflie build system. 

> That will be pretty bad. Your controller is called from the stabilizer loop (runs at 1000Hz) and your controller will block the stabilizer loop until your function returns.

Does that mean my controller update function has to be finished within 1000Hz, regardless of my controller update frequency? So correct me if I am wrong: all tasks (state estimation, controller, power distribution...) must be done within 1 ms so that they won't block the stabilizer loop.

I thought the firmware uses RTOS then blocking should not happen and the controller computing will be terminated at its predefined rate. 

I am implementing a single full-state LQR controller at 50-100Hz (but the computation is much more expensive). Do you have any comments or insights about that control structure? Is it even practical?

**Comment by krichardsson** [comment, marked as answer]
> My another problem is that I want to use Eigen. I am able to compile it for another STM32 with g++ (into C code). However, I cannot figure out how to do that with crazyflie build system.

In theory it should be possible to mix C and C++.[ There is an example app that shows this](https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/app_hello_world-cpp). I find that it often becomes complex and unfortunately we have very limited experience on this.

> Does that mean my controller update function has to be finished within 1000Hz, regardless of my controller update frequency? So correct me if I am wrong: all tasks (state estimation, controller, power distribution...) must be done within 1 ms so that they won't block the stabilizer loop.

Yes, but all hope is not lost, see below :-)

> I thought the firmware uses RTOS then blocking should not happen and the controller computing will be terminated at its predefined rate.

Sort of, ish. The OS will switch between code that is running in two different tasks (it will not be terminated, it will just be paused a bit) but in this case they are all running in the same task. Roughly something like this:
```C
// Simplified stabilizer loop
while(true) {
  // Lock until sensor data is available, every millisecond
  sensorsWaitDataReady();
  sensorsAcquire(&sensorData);
  // Call the estimator
  stateEstimator(&state, stabilizerStep);
  // Call the controller 
  controller(&control, &setpoint, &sensorData, &state, stabilizerStep);
  // Control the motors
  controlMotors(&control);
} 
```  
As you can see, if the estimator or controller takes too long we will not be ready in time for the next set of sensor data.

The solution to your problems is probably to run your controller in another task. The call to your controller from the stabilizer loop, is done at 1000Hz, and this should return quickly. The actual controller work is done in another task (that can be preempted by the stabilizer loop) and each iteration can take longer than 1 ms. The job of the (quick) function that is called from the stabilizer loop is to pass data back and forth between the stabilizer loop and the task. Note that the function that is called from the stabilizer loop is running in the context of the stabilizer task.

There is an example of a similar solution in the kalman estimator. The [`estimatorKalman()`](https://github.com/bitcraze/crazyflie-firmware/blob/cb1a5fa5064438a32193d2bcbc301c0b0068d4c6/src/modules/src/estimator/estimator_kalman.c#L276-L286) is called from the stabilizer loop and is essentially only copying a small chunk of memory. You can also see that a semaphore is used to protect the data from concurrency problems. 
The actual estimation work is done in the [`kalmanTask()` function](https://github.com/bitcraze/crazyflie-firmware/blob/cb1a5fa5064438a32193d2bcbc301c0b0068d4c6/src/modules/src/estimator/estimator_kalman.c#L207-L274). The main loop runs at 1000Hz (if possible) while the prediction part, that takes a bit longer, runs at 100Hz. The important thing is that the estimator is decoupled from the stabilizer loop. You can maybe get some inspiration from this solution?
